File: web/js/state.js
// state.js - Texas Hold'Em Lava Dome game state

class GameState {
    constructor() {
        this.reset();
    }

    reset() {
        // ── Session-level state ──────────────────────────────
        this.chips       = STARTING_CHIPS; // Active betting stack
        this.bank        = 0;              // Permanent cashed-out score
        this.round       = 0;              // Current round number (1-based when playing)
        this.sessionOver = false;
        this.escaped     = false;          // True if player chose to escape vs bust

        // ── Round-level state ────────────────────────────────
        this.phase       = 'idle';         // idle | betting | flop | turn | river | resolve | cashout
        this.ante        = 0;              // Ante paid this round
        this.currentBet  = 0;             // Total bet committed this round
        this.potWin      = 0;              // Chips won/lost last round resolution

        // ── Cards ────────────────────────────────────────────
        this.holeCards      = [];          // Player's 2 private cards
        this.communityCards = [];          // Up to 5 shared cards [flop x3, turn, river]
        this.deck           = null;        // Active Deck instance (set by Dealer)

        // ── Hand result ──────────────────────────────────────
        this.bestHand       = null;        // { name, cards, points } from last evaluation
        this.domeThreshold  = 0;           // Points needed to beat the dome this round
        this.beatDome       = false;       // Did the player beat the dome?

        // High scores are loaded once from JSONBin — not cleared on reset
    }

    // ── Derived getters ──────────────────────────────────────

    get totalWealth() {
        return this.chips + this.bank;
    }

    get currentAnte() {
        const idx = Math.min(this.round - 1, ANTE_SCHEDULE.length - 1);
        if (this.round - 1 < ANTE_SCHEDULE.length) {
            return ANTE_SCHEDULE[idx];
        }
        // Beyond schedule: last value + escalation per extra round
        const extra = (this.round - 1) - (ANTE_SCHEDULE.length - 1);
        return ANTE_SCHEDULE[ANTE_SCHEDULE.length - 1] + (extra * ANTE_ESCALATION_RATE);
    }

    get currentDomeThreshold() {
        return DOME_BASE_THRESHOLD + (this.round - 1) * DOME_THRESHOLD_SCALE;
    }

    get currentDepthLabel() {
        // Find the highest depth label whose round <= current round
        let label = DOME_DEPTHS[0];
        for (const depth of DOME_DEPTHS) {
            if (this.round >= depth.round) label = depth;
        }
        return label;
    }

    get allCards() {
        return [...this.holeCards, ...this.communityCards];
    }

    get canEscape() {
        return this.phase === 'cashout' && this.chips > 0;
    }

    get isBust() {
        return this.chips <= 0;
    }

    // ── Chip operations ──────────────────────────────────────

    cashOut(amount) {
        const actual = Math.min(amount, this.chips);
        this.chips -= actual;
        this.bank  += actual;
        return actual;
    }

    cashOutAll() {
        return this.cashOut(this.chips);
    }

    addChips(amount) {
        this.chips += amount;
    }

    removeChips(amount) {
        this.chips = Math.max(0, this.chips - amount);
        return this.chips === 0;
    }

    withdrawFromBank(amount) {
        const actual = Math.min(amount, this.bank);
        this.bank  -= actual;
        this.chips += actual;
        return actual;
    }

    // ── Serialization (for high score storage) ───────────────

    toScoreRecord(initials) {
        return {
            initials,
            bank:       this.bank,
            chips:      this.chips,
            totalScore: this.totalWealth,
            rounds:     this.round,
            escaped:    this.escaped
        };
    }
}

File: tui/lavadome/state.py
"""Session and round state.

A port of ``web/js/state.js``, which was already free of any browser
dependency — the whole of ``state``, ``config``, ``dome`` and ``betting`` is,
which is why this game was cheap to bring to a terminal.

Two chip pools, and the difference between them is the game: ``chips`` are in
play and can be lost, ``bank`` is cashed out and safe. The bank is the score.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from lavadome import config

#: idle → betting → flop → turn → river → resolve → cashout
PHASES = ("idle", "betting", "flop", "turn", "river", "resolve", "cashout")


@dataclass
class GameState:
    # ── Session ─────────────────────────────────────────────────────
    chips: int = config.STARTING_CHIPS   # active betting stack
    bank: int = 0                        # permanent cashed-out score
    round: int = 0                       # 1-based once playing
    session_over: bool = False
    escaped: bool = False                # left voluntarily, rather than bust

    # ── Round ───────────────────────────────────────────────────────
    phase: str = "idle"
    ante: int = 0
    current_bet: int = 0
    pot_win: int = 0                     # chips won (+) or lost (-) last round

    # ── Cards ───────────────────────────────────────────────────────
    hole_cards: list[Any] = field(default_factory=list)
    community_cards: list[Any] = field(default_factory=list)
    deck: Any = None

    # ── Result ──────────────────────────────────────────────────────
    best_hand: Any = None
    dome_threshold: int = 0              # fixed at round start, not re-derived
    beat_dome: bool = False

    def reset(self) -> None:
        self.chips = config.STARTING_CHIPS
        self.bank = 0
        self.round = 0
        self.session_over = False
        self.escaped = False
        self.phase = "idle"
        self.ante = 0
        self.current_bet = 0
        self.pot_win = 0
        self.hole_cards = []
        self.community_cards = []
        self.deck = None
        self.best_hand = None
        self.dome_threshold = 0
        self.beat_dome = False

    # ── Derived ─────────────────────────────────────────────────────

    @property
    def total_wealth(self) -> int:
        return self.chips + self.bank

    @property
    def current_ante(self) -> int:
        return config.ante_for_round(self.round)

    @property
    def current_dome_threshold(self) -> int:
        return config.threshold_for_round(self.round)

    @property
    def current_depth_label(self) -> str:
        return config.depth_label(self.round)

    @property
    def all_cards(self) -> list[Any]:
        return [*self.hole_cards, *self.community_cards]

    @property
    def can_escape(self) -> bool:
        return self.phase == "cashout" and self.chips > 0

    @property
    def is_bust(self) -> bool:
        return self.chips <= 0

    # ── Chips ───────────────────────────────────────────────────────

    def cash_out(self, amount: int) -> int:
        """Move chips to the bank. Returns how many actually moved."""
        actual = min(amount, self.chips)
        self.chips -= actual
        self.bank += actual
        return actual

    def cash_out_all(self) -> int:
        return self.cash_out(self.chips)

    def add_chips(self, amount: int) -> None:
        self.chips += amount

    def remove_chips(self, amount: int) -> bool:
        """Take chips off the stack. Returns True if that emptied it."""
        self.chips = max(0, self.chips - amount)
        return self.chips == 0

    def withdraw_from_bank(self, amount: int) -> int:
        """Put banked chips back in play — the only way money moves backwards."""
        actual = min(amount, self.bank)
        self.bank -= actual
        self.chips += actual
        return actual

    # ── Serialization ───────────────────────────────────────────────

    def to_score_record(self, initials: str) -> dict:
        return {
            "initials": initials,
            "bank": self.bank,
            "chips": self.chips,
            "totalScore": self.total_wealth,
            "rounds": self.round,
            "escaped": self.escaped,
        }
